### ISSUE TITLE
delete VPC resources

### ISSUE THREAD

**Issue by errordeveloper** [weaveworks/eksctl#103]
Currently deleting VPC stack will fail when there resources such as ELBs. We should be able to delete these, the questions is whether we should do it by default when we own the VPC or not?

**Comment by errordeveloper** [comment]
@JordanFaust indeed, full cleanup would be only done in the cases where VPC was created by us. Also, this is really a broad issue at this point – so for example deleting all cluster resource before delete the cluster itself could be sufficient in many cases.

**Comment by jstrachan** [comment]
maybe prompt the user or have a `--all` CLI option to delete everything? 

Its currently a huge PITA deleting clusters as usually there's something (usually a Load Balancer! sometimes other things...) causing the VPC to not get deleted resulting in lots of fun with the AWS console

**Comment by errordeveloper** [comment]
James, yes indeed, we would want a flag to control this, but I am leaning
towards having it should enabled by default (if VPC was one created by us).

**Comment by paulbsch** [comment]
This is of interest to me as I'm working with an app that creates k8s LoadBalancer services on demand.  So at the moment, it's not possible for me to cleanly delete the clusters I create with eksctl.  I think it makes sense for 'delete cluster' to automatically clean up the ELBs that were created with the eksctl created VPC.

**Comment by errordeveloper** [comment]
I think at present you should be able to delete the services first, then
wait for GC to kick in, but I do think this that is rather awkward and
waiting time may cost you.

**Comment by paulbsch** [comment]
I have been deleting the ELBs.  The VPC delete then gets hung up on the related security groups.  So it seems to require deleting those manually also.

**Comment by errordeveloper** [comment]
Paul, have you tried deleting service that own the ELB and waiting for some
time? I am pretty sure GC should kick in, but I just don't know the period
it runs at (will need to check).

**Comment by paulbsch** [comment]
I have done that.  The GC hasn't kicked in as I have seen.  At least it didn't do so in a short enough time frame.

**Comment by errordeveloper** [comment]
General info about handling stack deletion failures: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/troubleshooting.html#troubleshooting-errors-delete-stack-fails.

See also https://github.com/weaveworks/eksctl/pull/523#issuecomment-461623452.

**Comment by falfaro** [comment]
Any progress on this? Specifically on the ELB part?

**Comment by errordeveloper** [comment]
We have laid some ground work for this, there is now logic that deletes
stale ENIs, but more work is needed in this area before we can extend the
functionality.

**Comment by eladh** [comment]
Hi @errordeveloper  - is there any update ? 10x

**Comment by mplaul** [comment]
Hey @errordeveloper, any news on this issue?

**Comment by errordeveloper** [comment]
I did a simple test, and it looks ELBs certainly get deleted right away when a service is downgraded to `type: ClusterIP` (which is not always trivial, for example, I had to also clear `nodePort`).
```
$ kubectl describe service test
...
Events:
  Type    Reason                Age   From                Message
  ----    ------                ----  ----                -------
  Normal  EnsuringLoadBalancer  10m   service-controller  Ensuring load balancer
  Normal  EnsuredLoadBalancer   10m   service-controller  Ensured load balancer
  Normal  Type                  25s   service-controller  LoadBalancer -> ClusterIP
  Normal  DeletingLoadBalancer  25s   service-controller  Deleting load balancer
  Normal  DeletedLoadBalancer   14s   service-controller  Deleted load balancer
$ aws elb describe-load-balancers --region=us-west-2  --load-balancer-names=a06ca65f69f2c11e9abda02adb46d809 

An error occurred (LoadBalancerNotFound) when calling the DescribeLoadBalancers operation: There is no ACTIVE Load Balancer named 'a06ca65f69f2c11e9abda02adb46d809'
$
```

**Comment by 2opremio** [comment]
TL;DR: the code tells me the controller should always start the deletion of
an ELB mapped to a service within 30 seconds of the deletion of such
service (without needing to downgrade or modifying the service)

After reading the the release-1.13 code of
k8s.io/kubernetes/pkg/controller/service/service_controller.go , the
controller uses a 30-second period in its informer. So, the controller
should notice a service deletion in max 30 seconds.

Then, upon detecting a deletion, the controller calls
`EnsureLoadBalancerDeleted`, which (if EKS runs the upstream code at
k8s.io/pkg/cloudprovider/providers/aws/aws_loadbalancer.go ) should start
the deletion right away.

BTW, after reading `EnsureLoadBalancerDeleted` has taught me that it's
complicated enough not to try to replicate it ourselves.

So, if empirical evidence confirms what I have read I still think that
deleting the service and waiting for the mapped ELBs to disappear is our
best option.

PS: sorry for the lack of links, I am airborne and my Flight's internet
connection doesn't like GitHub.

**Comment by 2opremio** [comment]
The code I saw was at https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go

**Comment by rusyasoft** [comment]
I have created a cluster with new VPC (following to https://eksctl.io/usage/creating-and-managing-clusters/)
After successful creation I have run `eksctl delete cluster -f cluster.yaml` 

So far everything deleted except VPC, is it normal behavior? I'm just hoping when I apply eksctl to my production environment with my existing VPC, it won't wipe out my legacy VPC.

Is there any place where I can look-up for best practices integrating eksctl to a an existing production environment step-by-step ?

**Comment by michaelbeaumont** [comment]
@rusyasoft It only deletes VPCs that it created.
